icu4c/source/common/unistr.cpp

```diff
@@ -107,12 +109,36 @@ UOBJECT_DEFINE_RTTI_IMPLEMENTATION(UnicodeString)
 
 UnicodeString U_EXPORT2
 operator+ (const UnicodeString &s1, const UnicodeString &s2) {
-    return
-        UnicodeString(s1.length() + s2.length() + 1, static_cast<UChar32>(0), 0).
-            append(s1).
-                append(s2);
+  int32_t sumLengths;
+  if (uprv_add32_overflow(s1.length(), s2.length(), &sumLengths)) {
+    UnicodeString bogus;
+    bogus.setToBogus();
+    return bogus;
+  }
+  if (sumLengths != INT32_MAX) {
+    ++sumLengths;  // space for a terminating NUL if we need one
+  }
+  return UnicodeString(sumLengths, static_cast<UChar32>(0), 0).append(s1).append(s2);
 }
 
+U_COMMON_API UnicodeString U_EXPORT2
+unistr_internalConcat(const UnicodeString &s1, std::u16string_view s2) {
+  // TODO: version of uprv_add32_overflow() that takes a size_t for the second arg
+  // TODO: consider returning a std::optional<int32_t>
```

Comment:
I experimented with a slightly simpler version of that function for non-negative values where the second one is a size_t, but it didn't really make the total code simpler when it was effectively used in only one place.
```
namespace {

/** Version of uprv_add32_overflow() for non-negative numbers. */
std::optional<int32_t> addNonNegIntoInt32(int32_t a, size_t b) {
  U_ASSERT(a >= 0);
  if (b > INT32_MAX) {
    return {};
  }
  uint32_t sum = (uint32_t)a + (uint32_t)b;
  if (sum <= INT32_MAX) {
    return sum;
  } else {
    return {};
  }
}

}  // namespace
```